tapeout/src/main/scala/transforms/macros/MacroCompiler.scala

```diff
@@ -0,0 +1,401 @@
+// See LICENSE for license details.
+
+package barstools.tapeout.transforms.macros
+
+import firrtl._
+import firrtl.ir._
+import firrtl.PrimOps
+import firrtl.Utils._
+import firrtl.annotations._
+import firrtl.CompilerUtils.getLoweringTransforms
+import scala.collection.mutable.{ArrayBuffer, HashMap}
+import java.io.{File, FileWriter}
+import Utils._
+
+object MacroCompilerAnnotation {
+  def apply(c: String, mem: File, lib: Option[File], synflops: Boolean) = {
+    Annotation(CircuitName(c), classOf[MacroCompilerTransform],
+      s"${mem} %s ${synflops}".format(lib map (_.toString) getOrElse ""))
+  }
+  private val matcher = "([^ ]+) ([^ ]*) (true|false)".r
+  def unapply(a: Annotation) = a match {
+    case Annotation(CircuitName(c), t, matcher(mem, lib, synflops)) if t == classOf[MacroCompilerTransform] =>
+      Some((c, Some(new File(mem)), if (lib.isEmpty) None else Some(new File(lib)), synflops.toBoolean))
+    case _ => None
+  }
+}
+
+class MacroCompilerPass(mems: Option[Seq[Macro]],
+                        libs: Option[Seq[Macro]]) extends firrtl.passes.Pass {
+  def compile(mem: Macro, lib: Macro): Option[(Module, ExtModule)] = {
+    val pairedPorts = mem.sortedPorts zip lib.sortedPorts
+
+    // Parallel mapping
+    val pairs = ArrayBuffer[(BigInt, BigInt)]()
+    var last = 0
+    for (i <- 0 until mem.width.toInt) {
+      if (i <= last + 1) {
+        /* Palmer: Every memory is going to have to fit at least a single bit. */
+        // coninue
+      } else if ((i - last) % lib.width.toInt == 0) {
+        /* Palmer: It's possible that we rolled over a memory's width here,
+                   if so generate one. */
+        pairs += ((last, i-1))
+        last = i
+      } else {
+        /* Palmer: FIXME: This is a mess, I must just be super confused. */
+        for ((memPort, libPort) <- pairedPorts) {
+          (memPort.maskGran, libPort.maskGran) match {
+            case (_, Some(p)) if p == 1 => // continue
+            case (Some(p), _) if i % p == 0 =>
+              pairs += ((last, i-1))
+              last = i
+            case (_, None) => // continue
+            case (_, Some(p)) if p == lib.width => // continue
+            case _ =>
+              System.err println "Bit-mask (or unmasked) target memories are suppored only"
+              return None
+          }
+        }
+      }
+    }
+    pairs += ((last, mem.width.toInt - 1))
```

Comment:
`pairs` contains `(high, low)`s for bit extractions. It seems like bit extractions of PCAD IR must be `(high+1,low)`.